### DISCUSSION TITLE
Specify if input data is onehot for `KeepLargestConnectedComponent`

### DISCUSSION THREAD

**Discussion by function2-llx** [Project-MONAI/MONAI#3578] [Q&A]
**Is your feature request related to a problem? Please describe.**
I'm using the post-processing transform `KeepLargestConnectedComponent`, its calculation depends on if the input data is in onehot format. Currently, the transform determines if the input data is onehot by checking the number of channels.
https://github.com/Project-MONAI/MONAI/blob/f6648a3298ac31dfec7cda4064567c858c1aab1d/monai/transforms/post/array.py#L338

I think this may be inconvenient and error-prone. I'm working on different experiments with different numbers of segmentation categories and I'm always using onehot format. While some of them may only have one category, in this case, the current implementation will treat the input data as non-onehot. If I simply specify `applied_labels` with the number of channels, this would cause unexpected results.

**Describe the solution you'd like**
Add `onehot` parameter in `__init__` of `KeepLargestConnectedComponent` (e.g., just like what `DiceLoss` does). Consider making it required, or keep the current behavior as default and deprecate it gradually.

**Comment by Nic-Ma** [comment]
Hi @YuanTingHsieh ,

Could you please help share some comments here?
I feel maybe some misleading doc-string caused a wrong usage case here.

Thanks in advance.

**Comment by YuanTingHsieh** [comment]
For single channel data, you should pass `applied_labels=[1]`. (If 1 means foreground)

For multiple channel data, if the first channel is background, and you have 5 channels (categories) in total, you can pass `applied_labels=[1, 2, 3, 4]`

**Comment by function2-llx** [comment]
@YuanTingHsieh I can understand this by reading the doc. This is exactly what I mean by the inconvenience, because I'm always using onehot format no matter the data are single or multiple channels, and I'm not putting the background on the first channel. So I'll have to add an if clause when specifying `applied_labels` to check the data are single or multiple channels. I don't know if this is intended.

But it seems that if I put the background on the first channel, then it becomes consistent for both single and multiple channel data. Is this the preferred way to work with?

**Comment by YuanTingHsieh** [comment]
Hi @function2-llx 

Let's say your single channel data use pixel 1 as foreground.
Then its shape is [1, S1, S2, S3]
In this case you pass `applied_labels=[1]`.

Your multi-channel data has 3 channels and no background.
Then its shape is [3, S1, S2, S3]
In this case you pass `applied_labels=[0, 1, 2]` if you want all three channels to find their largest connected component.

This is the current usage.
What is the desire behavior that you are looking for?

**Comment by function2-llx** [comment]
@YuanTingHsieh Hello, sorry for the late reply.

I mean, I want to treat "single-channel data" as "data that can have variable channels only just has one channel". So in this case, I still want to pass `applied_labels=[0]` (in code, it may be something like `applied_labels=range(num_labels)`).

**Reply by Nic-Ma** [reply]
Hi @function2-llx , thanks for your explanation.
Now I got your point. Let me try to enhance the `KeepLargestConnectedComponent` transform for your suggestion soon, we also have a similar request from @holgerroth recently about slicing only specified channels of One-Hot data.

Thanks.

**Reply by Nic-Ma** [reply]
Hi @function2-llx ,

Submitted PR https://github.com/Project-MONAI/MONAI/pull/3647 for this ticket.

Thanks.

**Reply by function2-llx** [reply]
Glad to see this, thanks for your work!

**Comment by rijobro** [comment]
Just coming here from https://github.com/Project-MONAI/MONAI/pull/3776 (thanks @Nic-Ma). If something is one-hotted, then the sum along the channel dimension should be one. If an image is one-hotted, but only contains 1 channel, then I feel it should be filled with all 1s and be meaningless. If we're dealing with binary data, then the onehot will have 2 channels.

Perhaps I'm missing something, but is there a legitimate use case where the data is one-hotted but only has 1 channel dimension?

**Reply by wyli** [reply]
I had exactly the same question https://github.com/Project-MONAI/MONAI/pull/3647#discussion_r783048377 I think what they want is that when 'num_classes=2', a single channel input is enough, because the two channels of one-hot are complementary.

**Reply by Nic-Ma** [reply]
Hi @rijobro ,

I think the use case here is: for example, the One-Hot data has 2 channels, channel 0 is image, channel 1 is background, but some data of the dataset may only has the image channel.
@holgerroth also requested a feature several weeks ago to select expected channel from One-Hot output to compute other things.
So I think it's not robust to check One-Hot format only by channel numbers, it's better to have the option to set the flag, default to automatically check it by channel numbers (just like current dev implementation).

Thanks.

**Reply by wyli** [reply]
The only confusing case is the onehot/non-onehot for two-class segmentation. Perhaps having a dedicated flag "is_onehot" is overengineering from this perspective.
